File: edge.cpp
#include <cmath>
#include <stdexcept>

#include "edge.hpp"
#include "point.hpp"

edge::edge()
{
	throw std::runtime_error( "an edge must be initialized with  vertices" );
}

edge::edge( const point& one, const point& two ) :
	a( one ), b( two ), length( calc_length( one, two ) )
{
	if ( a == b )
		throw std::invalid_argument( "an edge must have distinct vertices" );
	if ( b < a ) {
		point t = a;
		a = b;
		b = t;
	}
}

edge::edge( const edge& that ) :
	a( that.a ), b( that.b ), length( that.length )
{
}

edge::~edge()
{
}

point edge::get_a( void ) const
{
	return a;
}

point edge::get_b( void ) const
{
	return b;
}

float edge::get_length( void ) const
{
	return this->length;
}

float edge::calc_length( const point& a, const point& b ) const
{
	int dx = a.get_x() - b.get_x();
	int dy = a.get_y() - b.get_y();
	return static_cast<float>(
			std::sqrt( static_cast<double>( dx * dx + dy * dy ) ) );
}

edge& edge::operator=( edge const& that )
{
	if ( this != &that ) {
		this->a = that.a;
		this->b = that.b;
		this->length = that.length;
	}
	return *this;
}

bool edge::operator==( const edge& that ) const
{
	return ( this->a == that.a ) && ( this->b == that.b );
}

bool edge::operator!=( const edge& that ) const
{
	return !(*this == that);
}

File: plot-java/mst/Edge.java
package mst;

import java.lang.Math;

/**
 * A graph edge.
 */
class Edge {
    /**
     * Calclulate the distance between points.
     * @param a one point
     * @param b another point
     * @return the straight-line distance from point a to point b
     */
    public static float distance(Point a, Point b) {
        int dx = a.getX() - b.getX();
        int dy = a.getY() - b.getY();
        return (float)Math.sqrt((double)(dx * dx + dy * dy));
    }

    /**
     * Construct an edge from two distinct points.
     * @param a a Point
     * @param b another Point
     */
    Edge(Point a, Point b) {
        if (a.equals(b)) {
            throw new IllegalArgumentException("an edge must have distinct vertices");
        } else if (a.lessThan(b)) {
            this.a = new Point(a);
            this.b = new Point(b);
        } else {
            this.a = new Point(b);
            this.b = new Point(a);
        }
        this.length = distance(a, b);
    }

    /**
     * Construct an edge as a copy of another edge.
     * @param e A Edge object.
     */
    Edge(Edge e) {
        this(e.getA(), e.getB());
    }

    /**
     * Get end point A.
     * @return A ref to Point A.
     */
    public Point getA() {
        return this.a;
    }

    /**
     * Get end point B.
     * @return A ref to Point B.
     */
    public Point getB() {
        return this.b;
    }

    /**
     * Get the length of the edge.
     * @return the distance between A and B
     */
    public float getLength() {
        return this.length;
    }

    /**
     * Compare edges for equality.
     * Edges are equal if they have the same vertices.
     * (a1 == a2 && b1 == b2 || a1 = b2 && b1 == a2 )
     */
    public boolean equals(Edge e) {
        return (this.a.equals(e.getA()) && this.b.equals(e.getB())) ||
               (this.a.equals(e.getB()) && this.b.equals(e.getB()));
    }

    //public edge& operator=( edge const& );
    //public bool operator!=( const edge& ) const;

    @SuppressWarnings("unused")
    private Edge() {} // An edge must be initialized with end points
    private Point a;
    private Point b;
    private float length;
}

